capstone/src/test/java/com/google/step/servlets/LoginServletTest.java

```diff
@@ -0,0 +1,125 @@
+package com.google.step.servlets;
+
+import static org.junit.Assert.*;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.repackaged.com.google.gson.JsonObject;
+import com.google.appengine.repackaged.com.google.gson.JsonParser;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.meterware.httpunit.GetMethodWebRequest;
+import com.meterware.httpunit.WebRequest;
+import com.meterware.httpunit.WebResponse;
+import com.meterware.servletunit.ServletRunner;
+import com.meterware.servletunit.ServletUnitClient;
+import java.io.IOException;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.xml.sax.SAXException;
+
+public class LoginServletTest {
+    private final LocalServiceTestHelper helper =
+            new LocalServiceTestHelper();
+
+    @Before
+    public void setUp() {
+        helper.setUp();
+    }
+
+    @After
+    public void tearDown() {
+        helper.tearDown();
+    }
+
+    @Test
+    public void testIsLoggedIn() {
+        // Check if this works to set the user as logged in and not the admin
+        helper.setEnvIsAdmin(false).setEnvIsLoggedIn(true);
+        UserService userService = UserServiceFactory.getUserService();
+        assertTrue(userService.isUserLoggedIn());
+        assertFalse(userService.isUserAdmin());
+    }
+
+    @Test
+    public void testIsNotLoggedIn() {
+        // Check if it works to show the user is not logged in
+        helper.setEnvIsAdmin(false).setEnvIsLoggedIn(false);
+        UserService userService = UserServiceFactory.getUserService();
+        assertFalse(userService.isUserLoggedIn());
+        // Can't also check if the user is an admin because the method will throw an error since the
+        // user is not even logged in.
+    }
+
+    @Test
+    public void testDoGetWithoutLogIn() throws IOException, SAXException {
+        // Testing doGet method without having the user logged in.
+        // Result: {"LoggedIn":"false","loginURL":"someURL"}
+
+        ServletRunner sr = new ServletRunner();
```

Comment:
Hmmm interesting. I would have thought something like this would suffice: https://stackoverflow.com/questions/5434419/how-to-test-my-servlet-using-junit

But This is cool too, if a little thicker than a unit test.

Reply:
I can look into using mockito for the future and then transferring these tests to mockito.